extlib/millhone/src/cmd/analyze_jars.rs

```diff
@@ -0,0 +1,188 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fs::File,
+    io::{BufWriter, Read},
+    path::PathBuf,
+};
+
+use clap::Parser;
+use fingerprint::Combined;
+use getset::Getters;
+use serde::{Deserialize, Serialize};
+use stable_eyre::{eyre::Context, Result};
+use tar::{Archive, Entry};
+use tracing::{debug, info, info_span, warn};
+use typed_builder::TypedBuilder;
+
+#[derive(Debug, Parser, Getters)]
+#[getset(get = "pub")]
+#[clap(version)]
+pub struct Subcommand {
+    /// The tar file image to search and fingerprint jars in.
+    image_tar_file: PathBuf,
+}
+
+const JAR_OBSERVATION: &str = "v1.discover.binary.jar";
+
+#[derive(Debug, PartialEq, Eq, Serialize, Clone)]
+struct DiscoveredJar {
+    kind: &'static str,
+    path: PathBuf,
+    fingerprints: Combined,
+}
+
+impl DiscoveredJar {
+    fn new(path: PathBuf, fingerprints: Combined) -> Self {
+        DiscoveredJar {
+            kind: JAR_OBSERVATION,
+            path,
+            fingerprints,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Deserialize)]
+struct OciManifest {
+    #[serde(rename = "Layers")]
+    layers: Vec<PathBuf>,
+}
+
+/// The path in the manifest file corresponding to a layer.
+#[derive(Debug, PartialEq, Eq, Serialize, Hash)]
+struct LayerPath(PathBuf);
+
+#[derive(Debug, PartialEq, Eq, Serialize, TypedBuilder)]
+struct JarAnalysis {
+    /// Jars and fingerprints associated with each layer in a jar file.
+    discovered_jars: HashMap<LayerPath, Vec<DiscoveredJar>>,
+}
+
+#[tracing::instrument]
+pub fn main(opts: Subcommand) -> Result<()> {
+    let tar_filename = opts.image_tar_file();
+    let jar_analysis = JarAnalysis {
+        discovered_jars: jars_in_container(opts.image_tar_file())
+            .with_context(|| format!("Analyze Jar Archive {:?}", tar_filename))?,
```

Comment:
[nit, optional] similarly i'd call this something like "analyze container"
also nitty and optional but remember that errors are commonly rendered like `error: {cause}: {cause}: ...`, so making them all lower case usually reads better
```suggestion
            .with_context(|| format!("analyze container: {:?}", tar_filename))?,
```